cron.py

```diff
@@ -0,0 +1,182 @@
+LM = "OpenAI" # language model
+CK = 30 # chunk size
+CT = 20 # cluster count
+DT = (datetime.now() - timedelta(hours=30)).date().strftime("%Y%m%d") # date
+LG = "English" # language
+
+THREAD_COUNT = 15
+
+def load_srt(id, lg):
+  lang = "" if lg == "Original" else ".en"
+  r = requests.get(f"{VISEXP}/{id}.transcript{lang}.srt")
+  r.raise_for_status()
+  return r.content
+
+
+def load_inventory(ch, dt, lg):
+  r = requests.get(f"{VISEXP}/{ch}.{dt}.inventory.json")
+  r.raise_for_status()
+  return pd.json_normalize(r.json(), record_path="shows").sort_values("start_time", ignore_index=True)
+
+
+def create_doc(txt, id, start, end):
+  return Document(page_content=txt, metadata={"id": id, "start": round(start.total_seconds()), "end": round(end.total_seconds())})
+
+
+def chunk_srt(sr, id, lim=3.0):
+  docs = []
+  ln = 0
+  txt = ""
+  start = end = timedelta()
+  for s in srt.parse(sr.decode()):
+    cl = (s.end - s.start).total_seconds()
+    if ln + cl > lim:
+      if txt:
+        docs.append(create_doc(txt, id, start, end))
+      ln = cl
+      txt = s.content
+      start = s.start
+      end = s.end
+    else:
+      ln += cl
+      txt += " " + s.content
+      end = s.end
+  if txt:
+    docs.append(create_doc(txt, id, start, end))
+  return docs
+
+
+def load_chunks(inventory, lg, ck):
+  chks = []
+  for i, r in inventory.iterrows():
+    try:
+      sr = load_srt(r.id, lg)
+    except HTTPError as _:
+      continue
+    chks += chunk_srt(sr, r.id, lim=ck)
+    if DEV_ENV: break;
+  return chks
+
+
+def load_vector(d):
+  embed = OpenAIEmbeddings()
+  result = embed.embed_query(d.page_content)
+  return result
+
+def select_docs(dt, ch, lg, lm, ck, ct):
+  print("loading chunks...")
+  docs = load_chunks(inventory, lg, ck)
+  docs_list = [(d,) for d in docs]
+
+  print("loading vectors...")
+  with ThreadPool(THREAD_COUNT) as pool:
+    vectors = pool.starmap(load_vector, docs_list)
+
+  print("number of vectors =", len(vectors))
+  kmeans = KMeans(n_clusters=ct, random_state=10, n_init=10).fit(vectors)
+  cent = sorted([np.argmin(np.linalg.norm(vectors - c, axis=1)) for c in kmeans.cluster_centers_])
+  return [docs[i] for i in cent]
+
+
+def id_to_time(id, start=0):
+  dt = IDDTRE.match(id).groups()[0]
+  return datetime.strptime(dt, "%Y%m%d_%H%M%S") + timedelta(seconds=start)
+
+
+def get_summary(d):
+  msg = f"""
+  ```{d}```
+
+  Create the most prominent headline from the text enclosed in three backticks (```) above, describe it in a paragraph, assign a category to it, determine whether it is of international interest, determine whether it is an advertisement, and assign the top three keywords in the following JSON format:
+
+  {{
+    "title": "<TITLE>",
+    "description": "<DESCRIPTION>",
+    "category": "<CATEGORY>",
+    "international_interest": true|false,
+    "advertisement": true|false,
+    "keywords": ["<KEYWORD1>", "<KEYWORD2>", "<KEYWORD3>"]
+  }}
+  """
+  res = openai.ChatCompletion.create(
+    model=MODEL,
```

Comment:
Should we have it as a dictionary to cater to the other LLMs?

Reply:
I agree with this as a next step, but can we push it through as is in order for the site to be usable for Mark to share with others?

Reply:
Done